mantidproject/mantid
# Reduce recursive includes in algorithm.h - first attempt

Fixes #15246 (well, one step). This

```
Algorithm.h: 155 includes
MatrixWorkspace.h: 138 includes
ExperimentInfo.h: 91 includes
```

has been reduced to:

```
Algorithm.h: 79 includes
MatrixWorkspace.h: 119 includes
ExperimentInfo.h: 69 includes
```

`Framework` build time has been reduced 4-5%.
Average number of (recursive) includes in `Framework` cpp files has been reduced by ~10%.
Further details in comments, #15246.

**To test:**
- there should not be any real code changes, just changes in includes and forward declarations.
- all platforms should be happy with the changes made.
- see that you agree with the direction this is heading.
- I tried to organize includes and use forward declares wherever possible.
- Further improvements are suggested in #15319. Feedback is very welcome.

After this, if an algorithm wants to use and manipulate MatrixWorkspace, Axis, Sample, or Run objects, their specific headers will need to be included. I find this neater, rather than having everything included in every algorithm. For now the speedup is not that great. We might get further speedups if we manage to hide better the heavily templatized and inlined headers like WorkspaceProperty.h, and the geometry stuff.
